price-estimator/src/filter.rs

```diff
@@ -91,6 +115,41 @@ async fn get_markets<T>(
     Ok(warp::reply::json(&result))
 }
 
+async fn estimate_sell_amount<T>(
+    token_pair: TokenPair,
+    price_in_quote: f64,
+    _query: QueryParameters,
+    price_rounding_buffer: f64,
+    orderbook: Arc<Orderbook<T>>,
+) -> Result<impl warp::Reply, Infallible> {
+    // NOTE: The price in quote is `sell_amount / buy_amount` which is the
+    // inverse of an exchange rate.
+    let exchange_rate = 1.0 / price_in_quote;
+    let transitive_order = orderbook
+        .get_pricegraph()
+        .await
+        .order_for_exchange_rate(token_pair.into(), exchange_rate);
+    let (buy_amount_in_base, sell_amount_in_quote) = transitive_order
+        .map(|order| {
+            (
+                apply_rounding_buffer(order.buy, price_rounding_buffer),
+                apply_rounding_buffer(order.sell, price_rounding_buffer),
```

Comment:
I don't fully understand this logic. We are applying the rounding buffer to both buy and sell amounts (why not just one?). Also, I think the way this is implemented now it will make both the buy and sell amounts smaller, which wouldn't really create a buffer, would it?

Reply:
So, my understanding was that the price rounding buffer was to make the exchange rate slightly more pessimistic to match the solver. Initially I had (incorrectly) thought that we just wanted to decrease the sell amount so that it is more pessimistic as well (and also decrease the buy amount by the same ratio so that the resulting order has exactly the quote price specified by the call).

Now that I think about it, maybe this is incorrect and we should be applying it to the input price so that the exchange rate is more pessimistic and restrictive (making it match potentially fewer orders). Then we can also apply this to the buy amount (as you suggested) which would make the input price exactly match the order exchange rate.

Reply:
That argument makes sense. Do I understand correctly that the ratio of buy/sell amount still doesn't always match the requested price? E.g. if I want to know how much DAI I can sell at a price of $1 and there is only one order in the book buying 100DAI at 2$, I would get sellAmount:100DAI and buyAmount:200$, right?

However, by rounding the price before computing the sell amount we ensure that the ratio is always strictly better than the requested price. If we only rounded the buy amount down, the response ratio could be slightly worse than the requested one.

Reply:
> Do I understand correctly that the ratio of buy/sell amount still doesn't always match the requested price?

This is true for the estimate, however, the route is implemented to return 100:100 though. This is something I wasn't sure about how we should implement it. Basically, from an exchange rate we can estimate two orders:

1. Best possible order where the buy and sell amounts are as large as possible such that `buy_amount / sell_amount >= xrate`.
2. Order at exchange rate where the sell amount is as large as possible, and `buy_amount / sell_amount == xrate`.

I opted for 2, since it more accurately expresses the willingness to pay, but this can be easily changed.

> If we only rounded the buy amount down, the response ratio could be slightly worse than the requested one.

Correct.